### ISSUE TITLE
Correct TCK License to be the SATCK and not Apache

### ISSUE THREAD

**Issue by brianoliver** [jsr107/jsr107tck#115]
The TCK license is currently incorrect (Apache 2.0).   It should be the Standalone TCK (SATCK) that is used for all TCKs across the Java Platform.

**Comment by ljacomet** [comment]
Is that really mandatory?
For example, bean validation has an Apache 2 licensed TCK. See http://beanvalidation.org/1.1/tck/

On the Ehcache side, we happily used the publicly available TCK to make sure we are compliant in our public builds for example.

**Comment by brianoliver** [comment]
Yes.  It's completely mandatory.  The JCache TCK was never approved or requested to be Apache 2.0.   Just because one JSR uses a particular license, doesn't mean JCache should also use that license.

The good news is that the TCK will remain public, but it must be licensed through the SATCK as all TCKs are.

**Comment by cruftex** [comment]
> The JCache TCK was never approved or requested to be Apache 2.0.

I am very confused. Just a quick check says, that the TCK 0.11 was published on 21-Oct-2013 to Maven central under Apache 2.0 license.

> ...but it must be licensed through the SATCK....

The current code base is licensed under Apache. This cannot be undone. At the moment every contributor to the TCK agrees that their submitted changes are licensed under Apache, because this is the effective license of the project.

The need to change the license of the project would IMHO only arise if there are significant contributions pending merging, that should not be Apache licensed. Which one?

If the TCK must be licensed (meaning: made available) under SATCK, then Oracle can just do so and publish the TCK under that license. Oracle has all rights to do so, this is covered by the JCP membership agreement. This doesn't mean that this project license needs change.

> The JCache TCK was never approved or requested to be Apache 2.0. 

If being asked, I think there would be many requests.

TL;DR:
Why cannot Oracle publish the TCK under the SATCK license? There is no change needed for that.

**Comment by brianoliver** [comment]
> I am very confused. Just a quick check says, that the TCK 0.11 was published on 21-Oct-2013 to
> Maven central under Apache 2.0 license.

That was a mistake.  We need to correct it.

> The current code base is licensed under Apache. 

That was a mistake.  We need to correct it.

> This cannot be undone. At the moment every contributor to the TCK agrees that their submitted 
> changes are licensed under Apache, because this is the effective license of the project.

Incorrect.  The only contributions made to the TCK are from those that have signed the OCA.  No other contributions have ever been permitted.

> The need to change the license of the project would IMHO only arise if there are significant
> contributions pending merging, that should not be Apache licensed. Which one?

There have virtually been no contributions thus far and all have occurred under OCA.

> If the TCK must be licensed (meaning: made available) under SATCK, then Oracle can just do so 
> and publish the TCK under that license. Oracle has all rights to do so, this is covered by the JCP 
> membership agreement. This doesn't mean that this project license needs change.

The TCK is only meant to be available under the SATCK license.   It was a mistake to do otherwise.  Any vendor wanting to claim compliance / compatibility with the JCache, via the TCK, must have executed the SATCK agreement.

The current situation was not the agreement reached between Oracle and Terracotta/Greg Luck.

> If being asked, I think there would be many requests.

We're not asking.  We're correcting the license (for the maintenance release) to that which was contractually agreed.  There should be no impact to any of the current licensees in terms of cost.

> Why cannot Oracle publish the TCK under the SATCK license? There is no change needed for
> that.

That's what we'll do.   The TCK download from the JCP will require an executed SATCK license (at zero cost).  That will include the source, compiled binary and user guide.  This way licensees know exactly the version to use for compliance testing.

We may keep the source on github public (waiting on confirmation on this), but implementations claiming TCK compliance will need to license and use the official TCK.

**Comment by cruftex** [comment]
>> This cannot be undone. At the moment every contributor to the TCK agrees that their submitted 
changes are licensed under Apache, because this is the effective license of the project.

> Incorrect. The only contributions made to the TCK are from those that have signed the OCA. No other contributions have ever been permitted.

Why incorrect? It's written in the OCA itself. Citing from the OCA:

> Any contribution we make available under any license will also be made available under a suitable FSF (Free Software Foundation) or OSI (Open Source Initiative) approved license.

See the full text here: http://www.oracle.com/technetwork/oca-405177.pdf

**Comment by brianoliver** [comment]
Oracle and other companies have contractual agreements that require the license to be SATCK.   I can't change this.   The current license is incorrect, was not approved and not agreed by the copyright owners (or Oracle).  There have been contractual objections to the current incorrect license that need to be resolved.

It doesn't matter what you or I think.  It needs to be corrected.   After that you/I/anyone can argue with their lawyers to get a new contract agreed, have the license change etc.

For now I'm trying my hardest to simply resolve this issue so no one ends up in court with zero cost to anyone / everyone.

**Comment by cruftex** [comment]
> There have been contractual objections to the current incorrect license that need to be resolved.

Does these objections/contracts say the TCK has to be available under SATCK *and* the TCK or parts of it *may not* be available as Open Source?

> For now I'm trying my hardest to simply resolve this issue....

It is now 3 years after this repository was published under Apache. Hundreds of people used, established projects and contributed to this under the the premise that it is Apache licensed.

I understand that the current situation shouldn't exist, it should have been resolved years ago, at least before JCache 1.0 was published. You are trying to correct it, which is no fun, and I agree, it has to be corrected.

The way you describe to fix it, means you fix a contractual agreement, but at the same time you break the agreement with others who worked under the premise that this project/artifact is Open Source. We already heard a voice from Terracotta, which is a major contributor. Maybe I am a romantic person, but still I think there can be a resolution that makes everyone happy.

**Comment by brianoliver** [comment]
> Does these objections/contracts say the TCK has to be available under SATCK and the TCK or 
> parts of it may not be available as Open Source?

Are you really asking me to disclose the contents of confidential information in negotiated contracts between licensees?   Seriously?

The agreements and approvals are very simple:   The TCK is to be licensed under SATCK.  End of story.

1. Greg and I screwed up.  
2. We're fixing it for JCache 1.1
3. It has no financial or legal impact on licensees.
4. It doesn't impact JCache 1.0

> Hundreds of people used, established projects and contributed to this under the the premise 
> that it is Apache licensed.

This is statement is complete and utter non-sense.  It's totally misleading to the community.

1. All contributions have been under the OCA.
2. There haven't been hundreds of people contributing to the TCK.

https://github.com/jsr107/jsr107tck/graphs/contributors

About 50% of the contributions were made by Greg Luck.  About 50% where made by Oracle employees.  The rest (three people) made less than 1% of the contributions and all under the OCA.

3. By definition, "people" use the TCK, they rarely if ever contribute to it.  The maintenance leads are the only people that can change it, based on objections from through the JCP.

Any contributions someone has made to a fork of the TCK, is not a contribution!

4. If you look at the JCP page for JCache, which is the official location for JCache information, you'll find that the license is correctly listed as SATCK.   It has always been this way.

https://jcp.org/en/jsr/detail?id=107

The problem is that Greg and I failed to update the source code to reflect that license.   

5. We're not changing the official license, as listed but the JCP on the official JCache page.  

We're fixing the source for JCache 1.1 TCK, which I've been told should not have been made public, but only made available available to licensees (with zero cost).

6. Testing a JCache implementation for compliance does not require an Apache license.

I applaud your intellectual efforts to "make everyone happy", but at the moment "everyone is not happy" and this very small change will guarantee everyone can happy, the licenses are consistent (they are currently contradictory), all with a zero cost impact.

**Comment by gregrluck** [comment]
Both licenses listed here are in conflict with the license headers on the RI And TCK source code files. The RI was always under an Apache 2 license as were the TCK tests. 

Most of these files and there headers were created before the license update of 5 March 2013.

**Comment by cruftex** [comment]
To move forward, here some resolution options/directions I can think of:

1. This TCK repository becomes private now. Links to the repository will break. The existing discussion in the TCK issues will be hard to understand, since the source is missing. The existing TCK user guide refers to this repository. No more compliance testing for now, confused users.

2. This TCK repository freezes and gets a prominent remark that this is not the latest and official TCK any more. Further development goes on in a private repository. Most likely this will result in forks of this repository at its current state.

3. This TCK repository keeps the Apache License and is getting additional updates by those contributors that like their contributions to be available as Open Source. It gets the prominent remark that this is not the latest and official TCK. The official TCK, that is licensed under the SATCK, is available only for download and includes additional tests, guides etc. that are only available under SATCK.

@brianoliver: This a brainstorming to the best of my knowledge and trying to be objective. If one of them is not complying to your contracts, that I didn't know, just rule it out.

Hope that helps.

**Comment by brianoliver** [comment]
@gregrluck 

>Both licenses listed here are in conflict with the license headers on the RI And TCK source code 
> files. The RI was always under an Apache 2 license as were the TCK tests.

1. We're not talking about the RI so why bring it up?  The RI is Apache 2.0 and everyone agrees with that.  There may be some header files (like in the API) that weren't correct, but they are now corrected.

2. Oracle Legal and JCP advise me that the TCK was never agreed to be Apache 2.0.   There was some earlier discussion (I was not involved) but going through several years of emails and the final contract show that it wasn't (as I've been told).

> Most of these files and there headers were created before the license update of 5 March 2013.

Yes.  And they don't match (what I've been told) with the final contracts etc.

**Comment by brianoliver** [comment]
@cruftex 
The resolution option is really simple.

1. Correct the license to that what was agreed (SATCK) for JCache 1.1
2. The community, I guess armed with some lawyers, formally request and negotiate with all of the copyright holders for a different license.

I don't own the copyright.  I'm not the person to negotiate with.  

> This TCK repository becomes private now. Links to the repository will break. The existing 
> discussion in the TCK issues will be hard to understand, since the source is missing. The existing 
> TCK user guide refers to this repository. No more compliance testing for now, confused users.

The TCK for JCache 1.1. becomes private to everyone except the licensees!   If you license it, you get it, for zero cost!   Hell, we could even give people github access!

> This TCK repository freezes and gets a prominent remark that this is not the latest and official 
> TCK any more. Further development goes on in a private repository. Most likely this will result in > forks of this repository at its current state.

What are you talking about?   JCache 1.1 won't freeze.   JCache 1.0 TCK will still be available via the JCP, like all TCKs.  We'll publish it there (as we were meant too).

> This TCK repository keeps the Apache License and is getting additional updates by those
> contributors that like their contributions to be available as Open Source. It gets the prominent 
> remark that this is not the latest and official TCK. The official TCK, that is licensed under the
> SATCK, is available only for download and includes additional tests, guides etc. that are only
> available under SATCK.

What are you talking about?   

1. The TCK can only be updated by the maintenance / spec leads.   There are *no* external contributions.  Random people can't contribute to and/or change the TCK as this would effectively break it for every vendor.   There's a JCP process for raising objections and dealing with TCK issues.

2. How can there be an "official" and "unofficial" TCK?   That makes no sense.

The TCK is that which is released by the JCP, for download.   Anything else, is not the TCK.

**Comment by ljacomet** [comment]
> We already heard a voice from Terracotta, which is a major contributor.

To be clear, I was asking a question in my own name and do not represent the view of Terracotta / Software AG on this specific point of licensing.

I have attracted attention to this issue internally so that if something needs to happen it will.

**Comment by brianoliver** [comment]
Thanks for the update @ljacomet (Louis).  Very much appreciated.

I've requested additional internal meetings to see how we can make progress.

**Comment by cruftex** [comment]
@brianoliver: 

Thanks for discussing this further. Since I see a lot questions coming up my way, maybe some reasoning of mine needs to be explained in more detail.

> What are you talking about?

> The TCK can only be updated by the maintenance / spec leads. There are no external contributions.

*Contributor* and *contributions* are legal terms as used for example in the "Oracle Contributor Agreement". The Contributor is the original copyright holder or the producer of the IP. Oracle(=You), Terracotta/Hazelcast(=Greg) and others are contributors in the legal meaning that I was talking about. A spec lead that is updating the TCK and inserting IP not of its own, is not a contributor of that IP.

Everything that was contributed in this repository is published under the Apache license. This was wrong, but changing the license of this now, will not affect the fact that: Everything that was contributed in this repository is published under the Apache license. If you change the license and publish the content elsewhere, then the legal effect of it is: It will *also* be available under that different license.

Changing the license and the visibility of this repository will have only an effect for upcoming contributions. At this moment three of my pull requests are pending. No other upcoming contributions are pending. I am fine with the fact that my contributions are published under Apache and SATCK or any other license that Oracle likes it to be.

Making this repository private, will not cure the wrong treatment of Oracle's IP or anybody else's that did not agree to the fact that it is Apache licensed.

I, as a contributor to this repository and therefore to TCK 1.1, have never signed any contract that says that my contributions will _only_ made available under SATCK. The OCA, that you are refering to, explicitly says otherwise. In this matter, I do not speak for anybody else.

If there are entities that want to contribute to the TCK now or in the future, but only want to make their contributions available under SATCK, they can do so, and they can make sure that their contributions are only made available via download, as you say. One possibility is to request that this repository becomes private, another possibility is to contribute only to the downloadable deliverable. However, if for example, Oracle only wants their contributions made available under SATCK, but is not contributing to the TCK now, there is no cause at all, since changes to the license would affect only new contributions.

If this repository becomes private, everybody who agrees to work in public and agrees to the fact that their contributions will be Apache and agrees that the rights are properly transferred as required by OCA and JCP to be able to be absorbed by the JSR107 TCK, can just do so at another place.

To be clear: I do not object against that this repository becomes private. But I do not like that it will impose additional burden on those who actively work on the TCK now and, on the other side, it will not cure the past contract breach for Oracle. If Oracle will do the major work on TCK 1.1, then it is a different story.

Sorry for the longish sentences, I know these are hard to digest, however, it's complex.

**Comment by brianoliver** [comment]
@cruftex 

Thanks for your response.   I'm no longer at liberty to comment on this matter.   I've raised the issue with Greg and our legal team to determine the next steps.

I have no idea how long this process will take, but based on previous experience, six months is a good guess.   In the mean time, I suspect JCache 1.1 won't be released and/or any new contributions will occur.

A terrible shame given the agreed TCK provides everything that is required, with zero cost.   The only cost now is a delay to the release and legal fees.

**Comment by gregrluck** [comment]
This is a very long thread. Like almost all JSRs the TCK is stated to be per https://jcp.org/en/jsr/detail?id=107:

TCK license
Commercial use
The TCK will be available for commercial use under this TCK license.

Non-Commercial use
As required by the Java Specification Participation Agreement (JSPA), the TCK will be licensed at no charge without support to qualified not-for-profit. The Compatibility Testing Scholarship Program will verify such qualification. Support may also be provided at no charge with approval of the scholarship board. For more information, please refer to:
http://www.oracle.com/technetwork/java/index-137188.html

The license section was changed 5 March 2013 to add CDDL and Classpath 2 for the RI.  The TCK license was not changed.

I have applied the license change. According to Brian it should be sufficient to do this without closing off the repo access. 

Practically because everyone with commit access has signed either the Oracle Contributor Agreement or the Terracotta Contributor Agreement, there are no changes to repository access.

For those using the TCK to implement an implementation, they now have the correct license in the source code header.

**Comment by cruftex** [comment]
Okay, that's a reasonable compromise.

> Practically because everyone with commit access has signed either the Oracle Contributor Agreement or the Terracotta Contributor Agreement, there are no changes to repository access.

Be careful: AFIAK the OCA only applies to Open Source projects, all contributors need to be JCP members.

**Comment by gregrluck** [comment]
To contribute to the spec and the TCK you also need to work for a company that has signed the JCP JSPA or you have signed it yourself.

As a JCP member, you give a license to the Spec Leads, which can then be sub-licensed.

4a) 
A. Contributions to the Spec Lead. You hereby grant to each Spec Lead (and, if different, Maintenance
Lead) for each Expert Group for which You are not the Spec Lead, with respect to the Output of the
JSR led by the Spec. Lead, a perpetual, non-exclusive, worldwide, royalty-free, fully paid-up and, subject to
Section 4.D, irrevocable, license, with

So I don't think we any IP problems here.
